### ISSUE TITLE
"search ." in resolv.conf can cause problems for clients (e.g., SSSD)

### ISSUE THREAD

**Issue by yrro** [systemd/systemd#28389]
### systemd version the issue has been seen with

253

### Used distribution

Fedora 38

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

When no DNS search domains are configured on any connection, there should be no `search` directive in `resolv.conf`.

(I don't believe other combinations of DHCP client & `resolv.conf` managers are in the habit of specifying `search .` when there are no search domains provided by the DHCP client to the `resolv.conf` manager.)

### Unexpected behaviour you saw

`search .` is present in `resolv.conf`. This causes [SSSD's dynamic DNS update process to fail](https://github.com/SSSD/sssd/issues/6322).

Yeah, this is SSSD's problem to deal with, but there's a chance this might cause other software to misbehave as well...

### Steps to reproduce the problem

If using DHCP, configure the server to not provide neither the domain name, not search list options. Restart networking (e.g., with NM, make sure there are no other connections active, then disconnect and reconnect the connection). Check `resolv.conf` for a `search .` line.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
SSSD logs `[nsupdate_get_addrs_done] (0x0040): Could not resolve address for this machine, error [5]: Input/output error, resolver returned: [8]: Misformatted domain name`
```

**Comment by poettering** [comment]
This has been like this for a longer time (3y+). I don#t think it makes sense to revert this now, and we do this for a good reason (to disable the search domain resolution from the local hostname)

Also, if we don#t change these things they will never be fixed properly in broken software.

Hence, I am strongly of the opinion that we should just close this, this is a bug in SSSD, and I am sure they'll fix it quickly (if they haven#t yet anyway). systemd is not the place to work around this.

Hence closing.

**Comment by yrro** [comment]
> to disable the search domain resolution from the local hostname

I did wonder what the point of `search .` was.

So some software does what, take the part of the kernel hostname and use _that_ as a search domain if there's no `search` line in `resolv.conf`? Yuck!

**Comment by yrro** [comment]
Memo to self: setting `ipv4.dns-search` to a FreeIPA domain on an NM connection puts `search ipa.example.com` into `resolv.conf`, but it also causes NM to push DNS info to `systemd-resolved`, resulting in the DNS servers from the connection being used for the domain in preference to `systemd-resolved`'s global configuration.

Setting `ipv4.ignore-auto-dns` to `yes` means that `ipv4.dns-search` is ignored, and `search .` gets pushed to `resolv.conf` breaking SSSD.

Setting `ipv4.dns-search` to something like `home.arpa` should solve my problem: `search home.arpa` in `resolv.conf` will fix dynamic DNS updates in SSSD, while `systemd-resolved`'s global configuration will be used for the FreeIPA domain.

(Alternatively, I could just tell NM to not push DNS settings to `systemd-resolved` at all).

**Comment by yrro** [comment]
A better workaround is to set `dns_resolver_use_search_list = False` in the relevant `domain` section of `sssd.conf`.

**Comment by yrro** [comment]
FYI, the bug that caused sssd to misbehave in the presence of `search .` was actually in the c-ares DNS resolver library and was fixed in version [1.20.0](https://c-ares.org/changelog.html#1_20_0).
